### PR TITLE
Refactor DbResults, DagFactory, Postgres, etc.

### PR DESCRIPTION
**Pull request by kevinji**
Created a few interfaces, converted Postgres to use String columns instead of int column indices, and other simplifications. This is the refactoring portion split off from #73.

### CODE REVIEW

File: src/main/java/edu/berkeley/ground/dao/models/cassandra/CassandraStructureVersionFactory.java [hunk 1]
```diff
@@ -115,25 +115,25 @@ public StructureVersion retrieveFromDatabase(long id) throws GroundException {
...
     List<DbDataContainer> attributePredicates = new ArrayList<>();
     attributePredicates.add(new DbDataContainer("structure_version_id", GroundType.LONG, id));
-    CassandraResults attributesSet = this.dbClient.equalitySelect("structure_version_attribute",
+
+    DbResults attributesSet = this.dbClient.equalitySelect("structure_version_attribute",
         DbClient.SELECT_STAR, attributePredicates);
 
     if (attributesSet.isEmpty()) {
-      throw new GroundException("No StructureVersion attributes found for id " + id + ".");
+      throw new GroundException("No attributes found for StructureVersion with id " + id + ".");
```
Comment (maasg) [inline src/main/java/edu/berkeley/ground/dao/models/cassandra/CassandraStructureVersionFactory.java]: maybe use `GroundVersionNotFoundException` instead?

**Review by maasg — COMMENTED** [review]
+1 for increasing the abstraction and removing the specific "<backend>clients" from the calling classes. 
What would you think of making `DBClient` an `Iterable` ?  That would make simple `for-expressions` possible, instead of `do...while` and `while...` loops.
Like here: https://github.com/ground-context/ground/pull/75/files#diff-bf7526f106f74a35c573404ebb058024L135

**Comment by kevinji** [comment]
That's a good idea.

**Comment by kevinji** [comment]
This should finally pass all tests! What an debugging experience.

**Comment by kevinji** [comment]
Just rebased this onto master, so this should be mergeable.

File: app/dao/models/cassandra/CassandraGraphVersionFactory.java [hunk 2]
```diff
@@ -132,16 +133,15 @@ public GraphVersion retrieveFromDatabase(long id) throws GroundException {
...
     List<Long> edgeVersionIds = new ArrayList<>();
     DbResults edgeSet = this.dbClient.equalitySelect("graph_version_edge",
         DbClient.SELECT_STAR, edgePredicate);
 
-    if (!edgeSet.isEmpty()) {
-      do {
-        edgeVersionIds.add(edgeSet.getLong("edge_version_id"));
-      } while (edgeSet.next());
+    for (DbRow edgeRow : edgeSet) {
+      edgeVersionIds.add(edgeRow.getLong("edge_version_id"));
```
Comment (maasg) [inline app/dao/models/cassandra/CassandraGraphVersionFactory.java]: :+1:

File: app/db/PostgresResults.java [hunk 3]
```diff
@@ -15,136 +15,194 @@
...
-   * @return false if there are no more rows
-   */
-  public boolean next() throws GroundDbException {
+    boolean isEmpty = true;
     try {
-      return this.resultSet.next();
+      // Move cursor to row 1.
+      isEmpty = !this.resultSet.next();
     } catch (SQLException e) {
-      throw new GroundDbException(e);
```
Comment (vsreekanti) [inline app/db/PostgresResults.java]: Why wouldn't we want to throw an exception here?

**Review by vsreekanti — APPROVED** [review]
This is awesome, @mc10 -- thanks!

File: app/db/PostgresResults.java (hunk 3 above)
Comment (kevinji) [reply app/db/PostgresResults.java]: I wasn't sure if constructors should throw an exception; I'll see if I can do that instead.

**Comment by vsreekanti** [comment]
Sorry for the wasted work, but `DBClient` is irrelevant not with the new Play architecture.

**Pull request closed without merging**